"Repeat group picker" UI for FormHierarchy

cooperka opened the pull request:
Implements the core behavior described in https://github.com/opendatakit/roadmap/issues/19. The new buttons will be added in a separate upcoming PR.

Based on branch https://github.com/opendatakit/collect/pull/2730.

![screenshot of new picker UI](https://user-images.githubusercontent.com/2047062/48794831-13d35a00-ecc9-11e8-900e-d2863e089244.png)

#### What has been done to verify that this works as intended?

Extensive manual testing with:
- nested-repeats-complex ([XML](https://drive.google.com/open?id=1zOvDEZz6vwQr5IBT91k7MuDMZ-hJDats) / [XLS](https://drive.google.com/open?id=1xVv682pbBMQ_MTZdNnj6aP11jhPp6fT8WQ42oEr7WaA)) (top-level "friends" and "enemies" groups; nested "friends/pets" group)
- nested-repeats-complex-jr ([XML](https://drive.google.com/open?id=1j6LPDAb9keqpinge-onLBtgrqBhEWdAy)) (same as above, but "enemies" has no `jr-template` tag, so it shows a blank child by default)

Navigating in and out of nested repeat groups, adding children, editing names, using the back button, jumping back and forth between the FormEditor and FormHierarchy views.

#### Why is this the best possible solution? Were any other approaches considered?

We considered overhauling `FormHierarchyActivity` to navigate using a more conventional "back stack" instead of using `refreshView`, but that requires all method calls to be stateless to avoid corrupting the state of earlier activities, which `FormController` currently doesn't support. This simpler solution works within the existing event-based form parsing framework.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

This changes the navigational UI so that instead of expanding a dropdown showing the children of a repeatable group, those children are shown on their own on an intermediary screen (sometimes referred to as the "picker"). This intermediary screen will be used in an upcoming PR to display an "add" button to add new items to the repeat group. Overall these changes aim to improve the speed and intuitiveness of navigation.

The behavior changes should be isolated to FormHierarchyActivity, but this activity interacts with the global state via FormController, so regressions can be avoided by ensuring that the state is not mutated in some new unexpected way.

#### Do we need any specific form for testing your changes? If so, please attach one.

Linked above.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.

I was unable to find any existing documentation that needs to be changed (namely [here](https://docs.opendatakit.org/collect-filling-forms/#jumping-to-questions) which has screenshots of FormHierarchy but no expandable repeat groups).

#### Before submitting this PR, please make sure you have:

- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

cooperka committed:
Factor out jumpToHierarchyStartIndex

cooperka committed:
Call refreshView instead of manual expand/collapse

This doesn't yet render anything new.

cooperka committed:
Set repeatGroupPickerRef in order to render 'picker' view

cooperka committed:
Handle 'go up' when in the picker view

cooperka committed:
Always step out to the closest screen event

This ensures a consistent form reference.

For example, without this change, if the user is on a question
then opens the jump screen, the index will be of the question
and not of the repeat group.

cooperka committed:
Toggle the picker if coming from an inner repeat group

cooperka committed:
Clarify currentIndex docs

cooperka committed:
Also show groupPath for the root picker

Currently no text to display, but at least render the UI.

cooperka committed:
Safe repeatGroupPickerIndex instead of repeatGroupPickerRef

cooperka committed:
Show repeat picker 'label' in groupPath

cooperka committed:
Compare unindexed ref to support nested repeat groups

cooperka committed:
Replace dropdown icon with 'Repeatable Group' label

cooperka committed:
Remove now-obsolete code from HierarchyElement

cooperka committed:
Factor out shouldShowRepeatGroupPicker()

cooperka committed:
Add clarifying comments in goUpLevel()

cooperka committed:
Fix checks

Remove unused imports; add braces; remove unused drawables.

cooperka committed:
Use 'instance' instead of 'child' when possible

cooperka committed:
Clean up old comments

yanokwa commented:
@zestyping I'd appreciate a careful review of this PR if you have a moment!

zestyping commented:
@yanokwa Thanks for looping me in! I'm in Dar Es Salaam wrangling map data for a hospital at the moment, so I might not get to this before the call on Friday.  I can try to have a look at it on the weekend, if that's not too late.

grzesiek2010 commented:
I know keeping arrow icons doesn't make sense now as this pr removes expanding/collapsing but maybe we should add another icon to indicate it's a group we can open. This pr adds just a text `Repeatable Group` but to be honest such a group looks exactly the same like normal question (it could be a text question with `Repeatable Group` as an answer)

| Before  | This branch |
| ------------- | ------------- |
| ![screenshot_20181206-141632](https://user-images.githubusercontent.com/3276264/49586676-205ee000-f962-11e8-919b-9fa85e05ace8.png)  | ![screenshot_2018-12-06-14-16-14](https://user-images.githubusercontent.com/3276264/49586637-f9081300-f961-11e8-9720-15be22d3ef3b.png)

what do you think?

Maybe one from those:
https://material.io/tools/icons/?icon=repeat&style=baseline
https://material.io/tools/icons/?icon=group_work&style=baseline
https://material.io/tools/icons/?icon=more&style=baseline
https://material.io/tools/icons/?icon=group&style=baseline

cooperka commented:
@grzesiek2010 I completely agree, and we actually have that in mind for an upcoming PR. I didn't want to put too many separate changes in one PR but we do want to add icons to each type of question to visually represent the question/item type (e.g. repeatable group, date, text, image, etc.). Would you be okay with holding off on that for now?

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/logic/HierarchyElement.java:
```diff
@@ -124,18 +107,13 @@ public void addChild(HierarchyElement h) {
          * A question.
          */
-        QUESTION;
+        QUESTION
```
QUESTION is question it's clear but PARENT and CHILD after your changes in fact are:
PARENT - reapatable group
CHILD - single group or something like that

so I would expect better naming here because it was confusing for me at first glance.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/activities/FormHierarchyActivity.java:
```diff
@@ -198,7 +231,99 @@ private String getCurrentPath() {
+    }
+
+    private String getGroupRef(FormController formController) {
+        return formController.getFormIndex().getReference().toString(true);
```
You can call just toString() because it uses true by default then:
https://github.com/opendatakit/javarosa/blob/master/src/org/javarosa/core/model/instance/TreeReference.java#L472

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/activities/FormHierarchyActivity.java:
```diff
@@ -198,7 +231,99 @@ private String getCurrentPath() {
+    }
+
+    private String getParentGroupRef(FormController formController) {
+        return formController.getFormIndex().getReference().getParentRef().toString(true);
```
You can call just toString() because it uses true by default then:
https://github.com/opendatakit/javarosa/blob/master/src/org/javarosa/core/model/instance/TreeReference.java#L472

grzesiek2010 commented:
> Would you be okay with holding off on that for now?

I would prefer having that one with this pr. I don't know how long it will take to review/accept and merge another pr. We can release a new version in the middle so it would be better to add a group icon in this pr. It's not a big deal to do that so if you can please do.

grzesiek2010 commented on collect_app/src/main/res/values/strings.xml:
```diff
@@ -116,6 +116,7 @@
     <string name="quit_application">Exit %s</string>
     <string name="quit_entry">Save Form and Exit</string>
     <string name="refresh">Refresh</string>
+    <string name="repeatable_group_label">Repeatable Group</string>
```
Once you add an icon I think this additional text is not needed.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/activities/FormHierarchyActivity.java:
```diff
@@ -312,6 +407,9 @@ public void refreshView() {
 
                 switch (event) {
                     case FormEntryController.EVENT_QUESTION:
+                        if (shouldShowRepeatGroupPicker()) {
```
What case does it cover? I mean how is it possible that we have `EVENT_QUESTION` but repeat group should be displayed?

cooperka replied:
This switch statement is part of a loop that goes through the entire form to render every item in the hierarchy. It skips items outside the current repeat group, but otherwise, it will render every question and nested repeat instance.

This code prevents questions from being rendered inside the repeat group picker, ensuring that only `CHILD` elements (nested repeat instances) are displayed.

cooperka commented:
Re: replacing the "Repeatable Group" text with just an icon, here's a mockup of what we're soon proposing:

![form icons](https://user-images.githubusercontent.com/2047062/49672811-52dc0c00-fa3a-11e8-9e16-7b77835be544.png)

The specific icons aren't important right now (we can discuss those in the actual proposal), but we think having sub-text for each item makes it visually more appealing. It also makes it easier to tell which questions are unanswered (helps with "Does blank mean I didn't put an answer, or is it supposed to be blank?").

I like your suggestion of adding the ["repeat" icon](https://material.io/tools/icons/?icon=repeat&style=baseline); do you think we can keep the text as well?

cooperka committed:
Remove default toString arg

cooperka committed:
Rename obsolete PARENT and CHILD types

cooperka committed:
Add repeat icon to repeatable groups in hierarchy

yanokwa commented:
I think keeping the text is a good idea. More information about what a thing is is helpful. 

~~The breadcrumb feels a little weird because you click on a thing I don't see that thing in the header.~~

~~current behavior~~
~~[screenshot](https://user-images.githubusercontent.com/32369/49706605-b487ac00-fbdb-11e8-92fd-767ec0047947.png)~~

~~desired behavior~~
~~[screenshot](https://user-images.githubusercontent.com/32369/49706640-ded96980-fbdb-11e8-877d-bd85d0dba055.png)~~

~~Is that easy to add to the PR? If so, great. If not, I can file an issue to follow up with it.~~

**Update**: Looks like you've already thought of that at https://github.com/cooperka/collect/pull/4

Finally, do you have the XLS that generated the XML, might be helpful for the testers.

cooperka commented:
Good eye, yes the breadcrumbs will be significantly cleaned up in that upcoming PR. I also just updated the description to include the XLS.

grzesiek2010 commented:
@cooperka thanks!

It looks good to me. @zestyping please take a look when you are free.

grzesiek2010 commented:
@mmarciniak90 @kkrawczyk123
whoever is going to test this pr please remember that we use the edited hierarchy view in two places:
- filling new forms
- reviewing sent forms

Everything should work as before apart from repeatable groups which will be displayed in a new way.
Please test form with many questions/groups to be sure the list is scrolled to an appropriate place once you open it.
And also repeatable groups that contain ordinary groups inside.

lognaturel commented:
I'll admit I'm a bit disappointed to see the stateful approach remain but I also agree that this is a pragmatic approach. The commit history is really clean. I've only done a quick read through but for what it's worth, I don't have any concerns.

cooperka commented:
I agree @lognaturel, I still hope we can get there eventually. The alternative stateless/back-stack approach is a much bigger project than we originally anticipated 😞 It seems to me like something that needs up-front commitment to see it through **all the way** (one piece at a time) if it's going to be attempted, otherwise we'll create the [n+1 problem](https://xkcd.com/927/) for form navigation. Hopefully we or someone else will be able to get our hands dirty in the near future.

kkrawczyk123 commented:
@cooperka, I really like the new hierarchy view as it looks very pretty but I have noticed some issues:

- When I open the general hierarchy view and open one of groups Friends/Enemies and rotate the screen I am thrown back to the general hierarchy view and I have to open Friends/Enemies group one again. The same situation occurs with the nested group: when I rotate the screen after opening Pets group I am thrown to the screen where I can open Pets group again. 

- When I add more group members then fit to the device's screen (No matter if Friends/Enemies/Pets) the scroll initially shows  in the middle of the displayed page like on attached screenshot: 
![screenshot_2018-12-13-15-03-23](https://user-images.githubusercontent.com/33342488/49940595-3a5b6e00-fee0-11e8-92e6-9288b4793e67.png)
and I have to scroll up to see the page from the first group. 

@opendatakit-bot unlabel "needs testing"

cooperka commented:
Thank you @kkrawczyk123! I'll address those and get back to you.

cooperka committed:
Remember repeatGroupPickerIndex on screen rotation

Manually restoreInstanceState because onRestoreInstanceState
happens too late after onCreate.

cooperka committed:
Don't try to scroll anywhere when entering/exiting the picker

This logic was obsolete; probably related to old expand/collapse behavior.
setAdapter is also already called by refreshView, so no need for it here.

Standardize on 'break' instead of 'return' for consistency.

cooperka commented:
@kkrawczyk123 can you try again please? I just made a couple fixes (tested and verified locally).

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/activities/FormHierarchyActivity.java:
```diff
@@ -170,6 +174,18 @@ public void onCreate(Bundle savedInstanceState) {
+    }
+
+    protected void restoreInstanceState(Bundle state) {
+        if (state == null) {
```
Why not just 
```
if (state != null) {
    repeatGroupPickerIndex = (FormIndex) state.getSerializable(REPEAT_GROUP_PICKER_INDEX_KEY);
}
```
then you don't need that return.
Or even if you created a separate method you can call it only if `state != null`

cooperka replied:
I personally like to check edge cases first and perform default logic at the end rather than nesting default logic inside conditions. I find that often leads to more maintainable code. No strong preference, it's just habit. I'm happy to change it if you want.

grzesiek2010 replied:
Generally, I wouldn't ask you to fix it but usually, in the whole project, we follow my approach. We don't have any check code rule for that unfortunately so I'm trying to be like a static code analyzer :smile:

cooperka replied:
Gotcha, I'll try to remember! Updated.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/activities/FormHierarchyActivity.java:
```diff
@@ -170,6 +174,18 @@ public void onCreate(Bundle savedInstanceState) {
+        super.onSaveInstanceState(outState);
+    }
+
+    protected void restoreInstanceState(Bundle state) {
```
Should be private.

cooperka committed:
Fix restoreInstanceState protected => private

cooperka committed:
Reverse restoreInstanceState condition

kkrawczyk123 commented:
**Tested with success!!**
Confirmed that issues have been fixed.
Verified on Androids: 4.1, 4.2, 4.4, 5.1, 6.0, 7.0 ad 8.1.

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

yanokwa commented:
@cooperka Looking forward to the other PRs that build on this one!
